### DISCUSSION TITLE
So I made a fully async DbDataAdapter - but how can I test Update/UpdateAsync correctly?

### DISCUSSION THREAD

**Discussion by daiplusplus** [dotnet/runtime#50402] [Show and tell]
I recently forked the older `AsyncDataAdapter` project ( originally https://www.nuget.org/packages/AsyncDataAdapter/ and https://github.com/voloda/AsyncDataAdapter ) to my own and completed the original creator's unfinished work to fully support all synchronous and asynchronous operations on `DbDataAdapter` which is now available here:

* https://www.nuget.org/packages/Jehoel.AsyncDataAdapter/
* https://github.com/Jehoel/AsyncDataAdapter/

I've attempted to add full test coverage - or at least, as much test coverage, as I can - however I'm stuck w.r.t. testing `DbDataAdapter.Update` and my `AsyncDbDataAdapter.UpdateAsync` because the documentation and examples are rather lacking, even the current reference-source code in is repo at https://github.com/dotnet/runtime/tree/main/src/libraries/System.Data.Common/src/System/Data seems to have no meaningful or useful comments - which is much-needed because of the various older C# idioms being used ("older", as in, the code dates back from .NET Framework 1.1 in 2000-2004, so it predates generics, extension methods, exception filters, etc). For example, the various `Fill`, `FillSchema`, and `Update` (and my `FillAsync`, `FillSchemaAsync`, and `UpdateAsync`) methods use undocumented mutually-exclusive parameters (e.g. When a method has both `DataSet dataset` OR `DataTable datatable` as parameters, the runtime caller must provide arguments for only one, not both, parameters). There is also the cryptic `srcTable` parameter in the `Update` method. [The official documentation does](https://docs.microsoft.com/en-us/dotnet/api/system.data.common.dbdataadapter.update?view=net-5.0) not aid in understanding _what exactly_ it means when the source-code itself gives no clue.

I was surprised to see that this repo only includes the _most basic_ of unit-tests for `DbDataAdapter`[1] [2]. Surely Microsoft has the original thorough test suite for `DbDataAdapter` somewhere? Can you bring it out of the archives and post it somewhere so the rest of us can use it to ensure our implementations are correct?

[1] https://github.com/dotnet/runtime/blob/main/src/libraries/System.Data.Common/tests/System/Data/Common/DataAdapterTest.cs
[2] https://github.com/dotnet/runtime/blob/main/src/libraries/System.Data.Common/tests/System/Data/Common/DbDataAdapterTest.cs

**Comment by danmoseley** [comment]
@roji

**Reply by roji** [reply]
DataAdapter in general is to a large extent "archived" - this is why the code isn't up to date with the latest C# features and coding practices, and hasn't received async support (see #22109). And yeah, since DataAdapter can fill either a single DataTable or a DataSet with multiple DataTables, the internal code usually unifies the two, with mutually exclusive parameters - that's just how things are architected. The official documentation you linked to only documents public APIs (where either DataTable or DataSet is accepted, but not both), and does not attempt to document internal code.

I'm also aware that general test coverage for these classes is lacking - @danmoseley any idea if we have any additional testing that did not make it across from the old codebase?

**Comment by ChristineBoersen** [comment]
Very nice work Jehoel
I see you had to do some of what I did when I internally forked Voloda's version to generalize the DBDataAdapter underneath SqlDataAdapter from Voloda's code.  

Also, you were right he had missed a few ConfigureAwait(false) and at least one or two fetchasync opportunities (just taking a quick glance at what I did to get Voloda's code working locally)
